File: Top 100 Codes/01-Getting Started/26-Harshad number/Harshadnumber.cpp
#include <iostream>
using namespace std;

// Function to check if a number is a Harshad (Niven) number
int checkHarshad(int num) {
    int sum = 0;      // Variable to store the sum of digits
    int temp = num;   // Temporary variable to process the number

    // Calculate the sum of digits of the number
    while (temp != 0) {
        sum += temp % 10; // Add the last digit to sum
        temp /= 10;       // Remove the last digit
    }

    // Check if the number is divisible by the sum of its digits
    return num % sum == 0;
}

int main() {
    int n;

    // Take user input
    cout << "Enter a number to check if it is a Harshad number: ";
    cin >> n;

    // Check if the number is a Harshad number and display the result
    if (checkHarshad(n))
        cout << n << " is a Harshad number." << endl;
    else
        cout << n << " is not a Harshad number." << endl;

    return 0;
}

// Time complexity: O(d)
// Space complexity: O(1)
// Where d is the number of digits in the input number

File: Top 100 Codes/01-Getting Started/26-Harshad number/Harshadnumber.py
# Function to check if a number is a Harshad (Niven) number
def is_harshad_number(num):
    # Store the original number
    original_num = num
    # Initialize the sum of digits
    digit_sum = 0

    # Calculate the sum of digits
    while num > 0:
        digit_sum += num % 10  # Add the last digit to the sum
        num //= 10            # Remove the last digit

    # Check if the original number is divisible by the sum of its digits
    return original_num % digit_sum == 0

# Take user input
n = int(input("Enter a number to check if it is a Harshad number: "))

# Check and print the result
if is_harshad_number(n):
    print(f"{n} is a Harshad number.")
else:
    print(f"{n} is not a Harshad number.")
